File: backend/src/main/java/org/kpet/petshop/Controllers/OrderController.java
package org.kpet.petshop.Controllers;

import jakarta.persistence.EntityNotFoundException;
import org.kpet.petshop.Models.Order;
import org.kpet.petshop.Models.OrderItem;
import org.kpet.petshop.Repositories.OrderRepository;
import org.kpet.petshop.Services.OrderService;
import org.kpet.petshop.Util.CsvWriter;
import org.springframework.http.ContentDisposition;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.time.LocalDateTime;
import java.time.YearMonth;
import java.time.format.DateTimeFormatter;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * REST surface for the order workflow. createOrder() is called right after the
 * WhatsApp message is sent from the cart (see CartComponent.checkoutOnWhatsapp) —
 * it's fire-and-forget from the frontend's perspective: a failure here doesn't
 * block the sale since the WhatsApp message already went out.
 */
@RestController
@RequestMapping("/api/orders")
public class OrderController {

    private final OrderService orderService;
    private final OrderRepository orderRepository;

    public OrderController(OrderService orderService, OrderRepository orderRepository) {
        this.orderService = orderService;
        this.orderRepository = orderRepository;
    }

    // Admin only (see SecurityConfig). Pinned to GET specifically: a bare
    // @RequestMapping here would also answer POST/PUT/DELETE, which
    // SecurityConfig only locks to ADMIN for GET on this path.
    @GetMapping(value = "/findAll", produces = "application/json")
    public List<Order> getAllOrders() {
        return orderRepository.findAll();
    }

    private static final List<String> ORDER_CSV_HEADERS = List.of(
            "N° de pedido", "Fecha", "Estado", "Producto", "Cantidad",
            "Precio unitario (Bs.)", "Subtotal (Bs.)", "Total del pedido (Bs.)", "Comprobante enviado");

    private static final Map<String, String> STATUS_LABELS_ES = Map.of(
            Order.PENDING_CONFIRMATION, "Pendiente a confirmar",
            Order.CONFIRMED, "Confirmado",
            Order.COMPLETED, "Finalizado",
            Order.CANCELLED, "Cancelado"
    );

    // Admin only (see SecurityConfig — "/export" is a single path segment
    // under /api/orders/**, already covered by the existing ADMIN-only
    // GET .../"/api/orders/*" rule, same as /api/products/export). Always
    // the current calendar month — "descargar el historial del mes" — not a
    // date range picker, so this stays a one-click download.
    //
    // One row per order ITEM rather than cramming every product into one
    // cell — Excel can only split a cell into columns along one delimiter,
    // so a "2x Bofe; 1x Traquea" cell reads as a single blob, not separate
    // values. Semicolon-delimited (see CsvWriter) since a comma-delimited
    // file opens as one giant column A on Excel set to a Spanish locale, and
    // a UTF-8 BOM up front so accented headers/values don't show as mojibake
    // when Excel opens the file directly instead of through an import wizard.
    @GetMapping("/export")
    public ResponseEntity<String> exportOrders() {
        YearMonth currentMonth = YearMonth.now();
        LocalDateTime start = currentMonth.atDay(1).atStartOfDay();
        LocalDateTime end = currentMonth.atEndOfMonth().atTime(23, 59, 59);

        List<Order> orders = orderRepository.findAll().stream()
                .filter(o -> o.getCreatedAt() != null && !o.getCreatedAt().isBefore(start) && !o.getCreatedAt().isAfter(end))
                .sorted(Comparator.comparing(Order::getCreatedAt))
                .toList();

        DateTimeFormatter dateFormat = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm");
        StringBuilder csv = new StringBuilder();
        csv.append('﻿');
        csv.append(CsvWriter.row(';', ORDER_CSV_HEADERS.toArray(new String[0]))).append("\n");

        for (Order o : orders) {
            String fecha = o.getCreatedAt().format(dateFormat);
            String estado = STATUS_LABELS_ES.getOrDefault(o.getStatus(), o.getStatus());
            String comprobante = o.isReceiptSent() ? "Sí" : "No";
            String total = formatAmount(o.getTotal());

            if (o.getItems().isEmpty()) {
                csv.append(CsvWriter.row(';',
                        String.valueOf(o.getId()), fecha, estado, "", "", "", "", total, comprobante
                )).append("\n");
                continue;
            }

            for (OrderItem item : o.getItems()) {
                double unitPrice = item.getUnitPrice() != null ? item.getUnitPrice() : 0;
                int quantity = item.getQuantity() != null ? item.getQuantity() : 0;
                csv.append(CsvWriter.row(';',
                        String.valueOf(o.getId()),
                        fecha,
                        estado,
                        item.getProductName(),
                        String.valueOf(quantity),
                        formatAmount(unitPrice),
                        formatAmount(unitPrice * quantity),
                        total,
                        comprobante
                )).append("\n");
            }
        }

        HttpHeaders headers = new HttpHeaders();
        headers.setContentDisposition(ContentDisposition.attachment()
                .filename("pedidos-" + currentMonth + ".csv").build());
        return ResponseEntity.ok()
                .headers(headers)
                .contentType(MediaType.parseMediaType("text/csv; charset=UTF-8"))
                .body(csv.toString());
    }

    private String formatAmount(double amount) {
        return String.format(Locale.US, "%.2f", amount);
    }

    @GetMapping("/{id}")
    public ResponseEntity<Order> getOrderById(@PathVariable Long id) {
        return orderService.getOrderById(id)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    // A logged-in customer's own order history ("Mis pedidos" in Mi Perfil).
    // Authenticated-only (see SecurityConfig) — every Client sees just their
    // own orders, scoped by the authenticated principal, never a client-supplied id.
    @GetMapping(value = "/mine", produces = "application/json")
    public List<Order> getMyOrders(Authentication authentication) {
        Long userId = Long.valueOf(authentication.getName());
        return orderRepository.findByUserIdOrderByCreatedAtDesc(userId);
    }

    // Public (guest checkout never needs an account — see SecurityConfig), but
    // if the caller happens to be logged in, the order is tagged with their id
    // so it shows up under "Mis pedidos" later. Never trusted from the request
    // body — always read from the authenticated principal, not client input.
    @RequestMapping(value = "/create", produces = "application/json", method = RequestMethod.POST)
    public ResponseEntity<?> createOrder(@RequestBody Order order, Authentication authentication) {
        if (authentication != null) {
            order.setUserId(Long.valueOf(authentication.getName()));
        }
        try {
            Order saved = orderService.createOrder(order);
            return ResponseEntity.created(URI.create("/api/orders/" + saved.getId())).body(saved);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    @RequestMapping(value = "/{id}/receipt-sent", produces = "application/json", method = RequestMethod.PUT)
    public ResponseEntity<Order> markReceiptSent(@PathVariable Long id) {
        try {
            return ResponseEntity.ok(orderService.markReceiptSent(id));
        } catch (EntityNotFoundException e) {
            return ResponseEntity.notFound().build();
        }
    }

    @RequestMapping(value = "/{id}/status", produces = "application/json", method = RequestMethod.PUT)
    public ResponseEntity<Order> updateStatus(@PathVariable Long id, @RequestBody Map<String, String> body) {
        try {
            return ResponseEntity.ok(orderService.updateStatus(id, body.get("status")));
        } catch (EntityNotFoundException e) {
            return ResponseEntity.notFound().build();
        }
    }
}

File: backend-php/app/Http/Controllers/OrderController.php
<?php

namespace App\Http\Controllers;

use App\Models\DiscountLead;
use App\Models\Order;
use App\Models\Product;
use App\Support\CsvWriter;
use Illuminate\Http\Request;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\DB;

class OrderController extends Controller
{
    public function findAll()
    {
        return response()->json(Order::orderByDesc('createdAt')->get());
    }

    private const STATUS_LABELS_ES = [
        Order::PENDING_CONFIRMATION => 'Pendiente a confirmar',
        Order::CONFIRMED => 'Confirmado',
        Order::COMPLETED => 'Finalizado',
        Order::CANCELLED => 'Cancelado',
    ];

    public function export()
    {
        $start = Carbon::now()->startOfMonth();
        $end = Carbon::now()->endOfMonth();
        $monthLabel = $start->format('Y-m');

        $orders = Order::whereBetween('createdAt', [$start, $end])->orderBy('createdAt')->get();

        $headers = ['N° de pedido', 'Fecha', 'Estado', 'Producto', 'Cantidad',
            'Precio unitario (Bs.)', 'Subtotal (Bs.)', 'Total del pedido (Bs.)', 'Comprobante enviado'];

        $csv = "\u{FEFF}" . CsvWriter::row($headers, ';') . "\n";

        foreach ($orders as $o) {
            $fecha = Carbon::parse($o->createdAt)->format('d/m/Y H:i');
            $estado = self::STATUS_LABELS_ES[$o->status] ?? $o->status;
            $comprobante = $o->receiptSent ? 'Sí' : 'No';
            $total = number_format($o->total, 2, '.', '');

            $items = $o->items;
            if (empty($items)) {
                $csv .= CsvWriter::row([$o->id, $fecha, $estado, '', '', '', '', $total, $comprobante], ';') . "\n";
                continue;
            }

            foreach ($items as $item) {
                $unitPrice = $item['unitPrice'] ?? 0;
                $quantity = $item['quantity'] ?? 0;
                $csv .= CsvWriter::row([
                    $o->id, $fecha, $estado, $item['productName'],
                    $quantity, number_format($unitPrice, 2, '.', ''),
                    number_format($unitPrice * $quantity, 2, '.', ''),
                    $total, $comprobante,
                ], ';') . "\n";
            }
        }

        return response($csv, 200, [
            'Content-Type' => 'text/csv; charset=UTF-8',
            'Content-Disposition' => "attachment; filename=\"pedidos-{$monthLabel}.csv\"",
        ]);
    }

    public function show($id)
    {
        $order = Order::find($id);
        if (!$order) {
            return response()->json(['error' => 'No encontrado'], 404);
        }
        return response()->json($order);
    }

    public function mine(Request $request)
    {
        $userId = $request->attributes->get('authUserId');
        return response()->json(Order::where('userId', $userId)->orderByDesc('createdAt')->get());
    }

    public function create(Request $request)
    {
        $data = $request->all();
        $items = $data['items'] ?? [];

        // Re-validated against live DB stock — never trusts the frontend
        // cart's cached snapshot, matching OrderServiceImpl.createOrder.
        foreach ($items as $item) {
            $productId = $item['productId'] ?? null;
            $quantity = $item['quantity'] ?? null;
            if ($productId === null || $quantity === null) {
                continue;
            }
            $product = Product::find($productId);
            if ($product && $product->stock !== null && $quantity > $product->stock) {
                return response()->json([
                    'error' => "No hay suficiente stock de \"{$product->name}\" (disponible: {$product->stock}).",
                ], 400);
            }
        }

        $userId = $request->attributes->get('authUserId');

        // The welcome-discount lead is only ever redeemed here — validated
        // server-side (not just trusted from the frontend) so a stale or
        // already-used id can't be replayed onto a second order.
        $discountPercent = null;
        $lead = null;
        $discountLeadId = $data['discountLeadId'] ?? null;
        if ($discountLeadId !== null) {
            $lead = DiscountLead::whereNull('redeemedAt')->find($discountLeadId);
            if ($lead) {
                $discountPercent = $lead->discountPercent;
            }
        }

        $order = Order::create([
            'total' => $data['total'] ?? 0,
            'status' => Order::PENDING_CONFIRMATION,
            'receiptSent' => false,
            'createdAt' => now(),
            'userId' => $userId, // never trusted from the request body
            'discountPercent' => $discountPercent,
        ]);
        $order->syncItems($items);

        if ($lead) {
            $lead->redeemedAt = now();
            $lead->orderId = $order->id;
            $lead->save();
        }

        $order->refresh();

        return response()->json($order, 201, ['Location' => "/api/orders/{$order->id}"]);
    }

    public function markReceiptSent($id)
    {
        $order = Order::find($id);
        if (!$order) {
            return response()->json(['error' => 'No encontrado'], 404);
        }
        $order->receiptSent = true;
        $order->save();
        return response()->json($order);
    }

    public function updateStatus(Request $request, $id)
    {
        $order = Order::find($id);
        if (!$order) {
            return response()->json(['error' => 'No encontrado'], 404);
        }

        $newStatus = $request->input('status');
        $oldStatus = $order->status;

        DB::transaction(function () use ($order, $oldStatus, $newStatus) {
            if ($oldStatus === Order::PENDING_CONFIRMATION && $newStatus === Order::CONFIRMED) {
                $this->deductStock($order->items);
            } elseif ($oldStatus === Order::CONFIRMED && $newStatus === Order::CANCELLED) {
                $this->restockItems($order->items);
            }
            $order->status = $newStatus;
            $order->save();
        });

        $order->refresh();
        return response()->json($order);
    }

    private function deductStock(array $items): void
    {
        foreach ($items as $item) {
            if (($item['productId'] ?? null) !== null && ($item['quantity'] ?? null) !== null) {
                Product::deductStock($item['productId'], $item['quantity']);
            }
        }
    }

    private function restockItems(array $items): void
    {
        foreach ($items as $item) {
            if (($item['productId'] ?? null) !== null && ($item['quantity'] ?? null) !== null) {
                Product::restockProduct($item['productId'], $item['quantity']);
            }
        }
    }
}
